src/wms/src/main/java/org/geoserver/wms/icons/IconProperties.java

```diff
@@ -130,54 +132,69 @@ public Double getHeading() {
+                            if (workspace != null) {
+                                workspaceStyles =
+                                        dataDir.getStyles(workspace).dir().getCanonicalFile();
+                            }
                             file = graphicFile = file.getCanonicalFile();
                             if (file.getAbsolutePath().startsWith(styles.getAbsolutePath())) {
                                 // ok, part of the styles directory, extract only the relative path
                                 String relativePath = file.getAbsolutePath()
                                         .substring(styles.getAbsolutePath().length() + 1);
                                 file = new File(relativePath);
-                            } else {
-                                // we wont' transform this, other dirs are not published
-                                file = null;
-                            }
-
-                            // rebuild the icon href accordingly
-                            if (file != null && styles != null) {
-                                return ResponseUtils.buildURL(
-                                        baseURL,
-                                        "styles/" + styles.toURI().relativize(graphicFile.toURI()),
-                                        null,
-                                        URLType.RESOURCE);
-                            } else {
-                                // we don't know how to handle this then...
-                                return null;
+                                if (file != null) {
+                                    return buildGraphicURL(
+                                            baseURL,
+                                            "styles",
+                                            styles.toURI()
+                                                    .relativize(graphicFile.toURI())
+                                                    .toString());
+                                }
+                            } else if (workspaceStyles != null
+                                    && file.getAbsolutePath().startsWith(workspaceStyles.getAbsolutePath())) {
+                                // part of the workspace styles directory
+                                String relativePath = file.getAbsolutePath()
+                                        .substring(workspaceStyles
+                                                        .getAbsolutePath()
+                                                        .length()
+                                                + 1);
+                                file = new File(relativePath);
+                                if (file != null) {
+                                    return buildGraphicURL(
+                                            baseURL,
+                                            "styles/" + workspace.getName(),
+                                            workspaceStyles
```

Comment:
Since you already rolled out a method for this, what about giving it both the styles and graphicfile reference, and handle the following business in there, rather than duplicating it twice?

```java
workspaceStyles
                                                    .toURI()
                                                    .relativize(graphicFile.toURI())
                                                    .toString()
```

Reply:
Good point